core/modules/replica/DatabaseServices.h

```diff
@@ -331,7 +558,156 @@ class DatabaseServices
      *   was not found in the configuration.
      */
     virtual size_t numOrphanChunks(std::string const& database,
-                           std::vector<std::string> const& uniqueOnWorkers) = 0;
+                                   std::vector<std::string> const& uniqueOnWorkers) = 0;
+
+    /**
+     * Log a Controller event
+     *
+     * @param event
+     *   event to be logged
+     */
+    virtual void logControllerEvent(ControllerEvent const& event) = 0;
+
+    /**
+     * Search the log of controller events for events in the specified time range.
+     *
+     * @param controllerId
+     *   unique identifier of a Controller whose events will be searched
+     *
+     * @param fromTimeStamp
+     *   (optional) the oldest (inclusive) timestamp for the search.
+     * 
+     * @param toTimeStamp
+     *   (optional) the most recent (inclusive) timestamp for the search.
+     *
+     * @param maxEntries
+     *   (optional) the maximum number of events to be reported. The default
+     *   values of 0 doesn't impose any limits.
+     *
+     * @return
+     *   collection of events found within the specified time interval
+     */
+    virtual std::list<ControllerEvent> readControllerEvents(std::string const& controllerId,
+                                                            uint64_t fromTimeStamp=0,
+                                                            uint64_t toTimeStamp=std::numeric_limits<uint64_t>::max(),
```

Comment:
nit: this line is too long (it makes the line wrapping weird in code review (I wish GitHub wouldn't wrap lines, and scrolled instead))

Reply:
Fixed